### ISSUE TITLE
Cannot forwarding same port from different pods

### ISSUE THREAD

**Issue by briandealwis** [GoogleContainerTools/skaffold#2660]
The _debug_ tests in `integration/testdata/debug` includes two NodeJS-based artifacts ([`nodejs`](https://github.com/GoogleContainerTools/skaffold/blob/master/integration/testdata/debug/nodejs/) and [`npm`](https://github.com/GoogleContainerTools/skaffold/blob/master/integration/testdata/debug/npm/)) that both expose port 3000 (HTTP) and 9229 (DevTools, exposed by transforms from `skaffold debug`).  But the port-forwards for the first one running are terminated when the same ports are forwarded for the second container.

Here is a filtered log.  In this case the `nodejs` artifact runs first.  When `npm` starts, the attempt to port-forward fails and results in `nodejs`'s port-forward being terminated.
```
$ cd integration/testdata/debug
$ .../skaffold debug -v debug --skip-tests --cache-artifacts=true --port-forward
[...]
Port forwarding pod/nodejs in namespace default, remote port 3000 -> local port 3000
INFO[0004] Streaming logs from pod: nodejs container: web 
Port forwarding pod/nodejs in namespace default, remote port 9229 -> local port 9229
DEBU[0004] Running command: [kubectl --context minikube logs --since=2s -f nodejs -c web --namespace default] 
[nodejs web] Debugger listening on ws://127.0.0.1:9229/d49ca909-55ae-4c65-8f25-6977a38c76e7
[nodejs web] For help, see: https://nodejs.org/en/docs/inspector
[nodejs web] Example app listening on port 3000!
INFO[0005] Streaming logs from pod: npm container: web  
DEBU[0005] Running command: [kubectl --context minikube logs --since=3s -f npm -c web --namespace default] 
DEBU[0005] Terminating port-forward pod-nodejs-default-3000 
Port forwarding pod/npm in namespace default, remote port 3000 -> local port 3000
failed to port forward pod-npm-default-3000, port 3000 is taken, retrying...
DEBU[0005] Terminating port-forward pod-nodejs-default-9229 
Port forwarding pod/npm in namespace default, remote port 9229 -> local port 9229
failed to port forward pod-npm-default-9229, port 9229 is taken, retrying...
DEBU[0005] terminated pod-nodejs-default-3000 due to context cancellation 
DEBU[0005] terminated pod-nodejs-default-9229 due to context cancellation 
[npm web] 
[npm web] > backend@1.0.0 start /opt/backend
[npm web] > node src/index.js
[npm web] 
[npm web] Debugger listening on ws://127.0.0.1:9229/554a11a7-dc3f-427d-93cd-217d38cb2f73
[npm web] For help, see: https://nodejs.org/en/docs/inspector
[npm web] Example app listening on port 3000!
[...]
port forwarding pod-npm-default-9229 recovered on port 9229
port forwarding pod-npm-default-3000 recovered on port 3000
```

**Comment by briandealwis** [comment]
The issue seems to be with `portForwardEntry.key()`:
https://github.com/GoogleContainerTools/skaffold/blob/1f086abda901338102e07d14199f26b92f873dbd/pkg/skaffold/kubernetes/portforward/port_forward_entry.go#L54-L62

In this case, the container names are the same.  Perhaps this code needs to instead trace the owner-reference links back to the defining resource (e.g., the pod's deployment name)?

**Comment by balopat** [comment]
We should fix this asap. 
The problem is not the key - `pod-nodejs-default-3000 != pod-npm-default-3000` - the deployment name _is_ in there. 
The problem is the port brokering logic before we start the dev loop: The desired behavior would be to broker a new port for the clashing one.

**Comment by briandealwis** [comment]
Two things to note:

  1. The log messages use `portForwardEntry.String()` and not `key()`, so the actual values used may not coincide with what is logged.
  2. This occurs with `skaffold debug` which sets `opts.PortForward.ForwardPods = true`, which I think enables `p.automaticPodForwarding`.

**Comment by balopat** [comment]
Hmmm...yeah I did get confused by `String()` and `key()`, the key might be the problem then.

**Comment by priyawadhwa** [comment]
So the issue is that `key()` doesn't include `podName` for pods, but this is by design so that when pods are recreated under different names, they are still forwarded to the same port.

Chatted with @balopat and we decided to include the top level owner of the object in the key which should solve this issue. I'll be working on this over the next couple days!

**Comment by balopat** [comment]
Update: the fix is coming in #2675.
